base/broadcast.jl

```diff
@@ -3,35 +3,64 @@
 module Broadcast
 
 using Base.Cartesian
-using Base: promote_op, promote_eltype, promote_eltype_op, @get!, _msk_end, unsafe_bitgetindex, linearindices, tail, OneTo, to_shape
+using Base: promote_eltype_op, @get!, _msk_end, unsafe_bitgetindex, linearindices, tail, OneTo, to_shape
 import Base: .+, .-, .*, ./, .\, .//, .==, .<, .!=, .<=, .÷, .%, .<<, .>>, .^
 export broadcast, broadcast!, bitbroadcast
 export broadcast_getindex, broadcast_setindex!
 
 ## Broadcasting utilities ##
 
-# fallback routines for broadcasting with no arguments or with scalars
-# to just produce a scalar result:
+# fallback routines for broadcasting with zero or one arguments
 broadcast(f) = f()
+broadcast(f, a) = f(a)
+broadcast(f, a::Tuple) = map(f, a)
+broadcast(f, A::AbstractArray) = broadcast_t(f, promote_eltype_op(f, A), A)
+
+# special cases for numbers and tuples
 broadcast(f, x::Number...) = f(x...)
+broadcast{N}(f, a::NTuple{N}, bs::Vararg{NTuple{N}}) = map(f, a, bs...)
+broadcast(f, As::AbstractArray...) = broadcast_t(f, promote_eltype_op(f, As...), As...)
 
 # special cases for "X .= ..." (broadcast!) assignments
 broadcast!(::typeof(identity), X::AbstractArray, x::Number) = fill!(X, x)
 broadcast!(f, X::AbstractArray) = fill!(X, f())
 broadcast!(f, X::AbstractArray, x::Number...) = fill!(X, f(x...))
 function broadcast!{T,S,N}(::typeof(identity), x::AbstractArray{T,N}, y::AbstractArray{S,N})
-    check_broadcast_shape(size(x), size(y))
+    _check_broadcast_shape(size(x), size(y))
     copy!(x, y)
 end
 
+# Traits to decide the shape of the resulting array
+abstract ContainerType
+immutable NoContainer    <: ContainerType end
+immutable TupleContainer <: ContainerType end
+immutable ArrayContainer <: ContainerType end
+
+containertype(x) = containertype(typeof(x))
+containertype(::Type) = NoContainer()
+containertype{T<:Tuple}(::Type{T}) = TupleContainer()
+containertype{T<:AbstractArray}(::Type{T}) = ArrayContainer()
+containertype(ct1, ct2) = and_containertype(containertype(ct1), containertype(ct2))
+containertype(ct1, ct2, cts...) = and_containertype(containertype(ct1), containertype(ct2, cts...))
+
+and_containertype(::ArrayContainer, ::ArrayContainer) = ArrayContainer()
+and_containertype(::ArrayContainer, ct) = ArrayContainer()
+and_containertype(ct, ::ArrayContainer) = ArrayContainer()
+and_containertype(::TupleContainer, ::TupleContainer) = TupleContainer()
+and_containertype(::TupleContainer, ::NoContainer) = TupleContainer()
+and_containertype(::NoContainer, ::TupleContainer) = TupleContainer()
+and_containertype(ct1, ct2) = NoContainer()
+
 ## Calculate the broadcast shape of the arguments, or error if incompatible
 # array inputs
 broadcast_shape() = ()
-broadcast_shape(A) = indices(A)
-@inline broadcast_shape(A, B...) = broadcast_shape((), indices(A), map(indices, B)...)
+broadcast_shape(A) = ()
+broadcast_shape(A::Tuple) = (OneTo(length(A)),)
+broadcast_shape(A::AbstractArray) = indices(A)
```

Comment:
I can change it, but in the context of the broadcast methods it is used to generate the "shape" of the resulting array, for which the name seems more appropriate.